src/ast/passes/semantic_analyser.cpp

```diff
@@ -665,7 +665,7 @@ void SemanticAnalyser::visit(Call &call)
       }
     }
 
-    Visit(call.vargs[i]);
+    call.vargs[i] = dereference_if_needed(call.vargs[i]);
```

Comment:
Is there a way to hide `dereference_if_needed()` in a more central place like before? Or do you think it's better to open code it like this?

Reply:
I will take a think on this when I rebase. Since this is potentially mutating the value, I like this to be explicit. But there probably needs to be a way for the visitor to mutate the AST during passes, so maybe I will need to tweak these interfaces.

Reply:
I've updated the pass and will push shortly, here's my thinking:

YES, but I don't want to do it in the pass. I've updated the visitor to allow for mutating the graph (e.g. define "visit" if you want to just walk the graph, and "replace" for the node type if you want to consider replacing it). I think we should try to organize the passes and just break this out into a separate "AutomaticDereference" pass (that could basically be defined for the generic expression and be a single method). My general feeling on a lot of the passes (esp. codegen, but that's another story) is that to the extent they can be made easier-to-understand & self-contained, the better. That's part of the reason for doing this Visitor change (so I can start adding passes that are very small and self-contained).

Reply:
Ack. +1 on smaller self-contained passes

Reply:
Closing on this, it would be a follow-up item.